third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -127,6 +127,35 @@ impl ScriptCache {
     }
 }
 
+#[derive(Clone)]
+pub(crate) struct StructNameCache {
+    idx_map: BTreeMap<StructIdentifier, StructNameIndex>,
+    names: Vec<StructIdentifier>,
+}
+
+impl StructNameCache {
+    pub(crate) fn new() -> Self {
+        Self {
+            idx_map: BTreeMap::new(),
+            names: vec![],
+        }
+    }
+
+    pub(crate) fn insert_or_get(&mut self, name: StructIdentifier) -> StructNameIndex {
+        if let Some(idx) = self.idx_map.get(&name) {
+            return *idx;
+        }
+        let idx = StructNameIndex(self.names.len());
+        self.idx_map.insert(name.clone(), idx);
+        self.names.push(name);
+        idx
+    }
+
+    pub(crate) fn fetch_idx(&self, idx: StructNameIndex) -> &StructIdentifier {
```

Comment:
`idx_to_name`?

Also, we are mixing `Identifier` and `Name` - let's be consistent.